src/poetry/core/masonry/builders/builder.py

```diff
@@ -378,14 +380,11 @@ def __init__(
 
         self.path = self.path.resolve()
 
-    def __eq__(self, other: BuildIncludeFile | Path) -> bool:
-        if hasattr(other, "path"):
-            return self.path == other.path
-
-        return self.path == other
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, BuildIncludeFile):
```

Comment:
Are we sure we were not relying on this for any contains check anywhere?

Reply:
Pretty sure: the tests are decent here, and use of the BuildIncludeFile is limited to `poetry.core.masonry.builders` so there aren't too many places to check.

I suppose there's always non-zero risk with a change; but for me part of the point of this initial round of typechecking is to find opportunities to clean up the code, and I think the pipelines provide a good level of confidence.

Reply:
An example of a trap that the previous implementation left for the unwary:

```
>>> foo = Path("a")
>>> bar = BuildIncludeFile(foo, "b")
>>> foo == bar
False
```
(this one is because the `BuildIncludeFile` makes its path absolute)

And here's a (perhaps less likely) example where this `__eq__` implementation gives non-transitive equality:
```python
#!/usr/bin/env python3

from pathlib import Path
from poetry.core.masonry.builders.builder import BuildIncludeFile

class HasPath:
    def __init__(self, path):
        self.path = path

a = Path("/foo/bar")
b = BuildIncludeFile(a, "whatever")
c = HasPath(a)

print(f"a == b ? {a == b}")
print(f"b == c ? {b == c}")
print(f"a == c ? {a == c}")
```
with output
```
a == b ? True
b == c ? True
a == c ? False
```